packages/aserto/src/aserto/client/directory/v2/__init__.py

```diff
@@ -19,23 +19,87 @@
 from aserto.client.directory.v2.helpers import ObjectIdentifier, RelationResponse
 
 
+def build_grpc_channel(address: str, default_address: str, ca_cert_path: str) -> Any:
+    if address == "" and default_address == "":
+        return None
+        
+    if address == "":
+        return grpc.secure_channel(
+        target=default_address, credentials=directory.channel_credentials(cert=ca_cert_path)
+    )
+        
+    return grpc.secure_channel(
+        target=address, credentials=directory.channel_credentials(cert=ca_cert_path)
+    )
+
 class Directory:
     def __init__(
         self,
         *,
-        address: str,
         api_key: str = "",
         tenant_id: str = "",
         ca_cert_path: str = "",
+        address: str = "",
+        reader_address: str = "",
+        writer_address: str = "",
+        importer_address: str = "",
+        exporter_address: str = "",
     ) -> None:
-        self._channel = grpc.secure_channel(
-            target=address, credentials=directory.channel_credentials(cert=ca_cert_path)
-        )
+        validation_error = directory.validate_addresses(address=address, reader_address=reader_address, writer_address=writer_address,
+                              importer_address=importer_address, exporter_address=exporter_address, model_address="")
+        if validation_error is not None:
+            raise validation_error
+
+        self._reader_channel = build_grpc_channel(reader_address, address, ca_cert_path)
+        self._writer_channel = build_grpc_channel(writer_address, address, ca_cert_path)
+        self._importer_channel = build_grpc_channel(importer_address, address, ca_cert_path)
+        self._exporter_channel = build_grpc_channel(exporter_address, address, ca_cert_path)
+
         self._metadata = directory.get_metadata(api_key=api_key, tenant_id=tenant_id)
-        self.reader = reader.ReaderStub(self._channel)
-        self.writer = writer.WriterStub(self._channel)
-        self.importer = importer.ImporterStub(self._channel)
-        self.exporter = exporter.ExporterStub(self._channel)
+        self._reader = (
+            reader.ReaderStub(self._reader_channel)
+            if self._reader_channel is not None
+            else None
+        )
+        self._writer = (
+            writer.WriterStub(self._writer_channel)
+            if self._writer_channel is not None
+            else None
+        )
+        self._importer = (
+            importer.ImporterStub(self._importer_channel)
+            if self._importer_channel is not None
+            else None
+        )
+        self._exporter = (
+            exporter.ExporterStub(self._exporter_channel)
+            if self._exporter_channel is not None
+            else None
+        )
+
+    def reader(self) -> reader.ReaderStub:
+        if self._reader is None:
+            raise directory.NilClient
```

Comment:
I wouldn't name this exception `NilClient` because:
1. 'Nil' isn't a thing in python. Python has `None` but no `nil`.
2. The fact that a client is None is a symptom, not the root cause.

I'd name this `ConfigError` instead and include the name of the unconfigured service in the exception message.